code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateImageUtils.kt

```diff
@@ -0,0 +1,311 @@
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.services.NetworkCallback
+import com.adobe.marketing.mobile.services.NetworkRequest
+import com.adobe.marketing.mobile.services.ServiceProvider
+import com.adobe.marketing.mobile.services.caching.CacheEntry
+import com.adobe.marketing.mobile.services.caching.CacheExpiry
+import com.adobe.marketing.mobile.services.caching.CacheService
+import com.adobe.marketing.mobile.util.UrlUtils
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.IOException
+import java.io.InputStream
+import java.net.HttpURLConnection
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Utility functions to assist in downloading and caching images for push template notifications.
+ */
+
+internal object PushTemplateImageUtils {
+    private const val SELF_TAG = "PushTemplateImageUtil"
+    private const val FULL_BITMAP_QUALITY = 100
+    private const val DOWNLOAD_TIMEOUT_SECS = 10
+
+    /**
+     * Downloads and caches images provided in the [urlList]. Prior to downloading, the image url
+     * is used to retrieve a [CacheResult] containing a previously cached image.
+     * If a valid cache result is returned then no image is downloaded.
+     * If no cache result is returned, a call to [downloadImage] is made to download then cache the image.
+     *
+     * This is a blocking method that returns only after the download for all images
+     * have finished either by failing or successfully downloading, or the timeout has been reached.
+     *
+     * @param cacheService the AEPSDK [CacheService] to use for caching downloaded image assets
+     * @param urlList [String] containing an image asset url
+     * @return [Int] number of images that were found in cache or successfully downloaded
+     */
+    internal fun cacheImages(
+        cacheService: CacheService,
+        urlList: List<String?>
+    ): Int {
+        val assetCacheLocation = getAssetCacheLocation()
+        if (urlList.isEmpty() || assetCacheLocation.isNullOrEmpty()) {
+            return 0
+        }
+
+        var downloadedImageCount = AtomicInteger(0)
+        val latch = CountDownLatch(urlList.size)
+        for (url in urlList) {
+            if (url == null || !UrlUtils.isValidUrl(url)) {
+                latch.countDown()
+                continue
+            }
+
+            val cacheResult = cacheService[assetCacheLocation, url]
+            if (cacheResult != null) {
+                Log.trace(
+                    PushTemplateConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Found cached image for $url"
+                )
+                downloadedImageCount.incrementAndGet()
+                latch.countDown()
+                continue
+            }
+
+            downloadImage(url) { image ->
+                image?.let {
```

Comment:
I don't think we need to cache the image if the latch times out or is interrupted since the notification UI might have already been built and displayed. Added a boolean flag to ignore the download result in this case.